src/main/io/gps.c

```diff
@@ -1576,57 +2089,83 @@ void _update_checksum(uint8_t *data, uint8_t len, uint8_t *ck_a, uint8_t *ck_b)
 static bool UBLOX_parse_gps(void)
 {
     uint32_t i;
+    const uint32_t weekDurationMs = 7 * 24 * 3600 * 1000;
+    int navDeltaTimeMs = 0;
 
     *gpsPacketLogChar = LOG_IGNORED;
-
-    switch (_msg_id) {
-    case MSG_POSLLH:
+#define CLSMSG(cls, msg) (((cls) << 8) | (msg))
+    switch (CLSMSG(_class, _msg_id)) {
+
+    case CLSMSG(CLASS_MON, MSG_MON_VER):
+        *gpsPacketLogChar = LOG_UBLOX_MONVER;
+        gpsData.platformVersion = ubloxParseVersion(strtoul(_buffer.ver.hwVersion, NULL, 16));
+        gpsData.ubloxM7orAbove = gpsData.platformVersion >= UBX_VERSION_M7;
+        gpsData.ubloxM8orAbove = gpsData.platformVersion >= UBX_VERSION_M8;
+        gpsData.ubloxM9orAbove = gpsData.platformVersion >= UBX_VERSION_M9;
+        gpsData.ackState = UBLOX_ACK_GOT_ACK;
+        break;
+    case CLSMSG(CLASS_NAV, MSG_NAV_POSLLH):
         *gpsPacketLogChar = LOG_UBLOX_POSLLH;
         //i2c_dataset.time                = _buffer.posllh.time;
         gpsSol.llh.lon = _buffer.posllh.longitude;
         gpsSol.llh.lat = _buffer.posllh.latitude;
         gpsSol.llh.altCm = _buffer.posllh.altitudeMslMm / 10;  //alt in cm
         gpsSol.time = _buffer.posllh.time;
+
+        // calculate the interval between nav packets, handling iTow wraparound at the end of the week
+        navDeltaTimeMs = (gpsSol.time - gpsData.lastNavSolTs + weekDurationMs) % weekDurationMs;
```

Comment:
`(x + a) % a` is mathematically the same as `(x) % a`, so 
`(gpsSol.time - gpsData.lastNavSolTs + weekDurationMs) % weekDurationMs`
should be the same as
`(gpsSol.time - gpsData.lastNavSolTs) % weekDurationMs`

Reply:
Wouldn't we then get the modulo of a very negative value?

Reply:
OK, I see, this is the internal iTOW field, and we use it for debug only anyway (?)
In that case it's probably okay not even to bother with modulo, just simply store `gpsSol.time - gpsData.lastNavSolTs` as not only it will be corrected during the next GPS reading anyway (in 100ms), but we also `constrain` the value between reasonable values so even if we are unlucky and one reading is before and the other is after the overflow, it won't matter to anyone anywhere.

Reply:
I'm glad you drew my attention to it.  Maybe it should be:
``navDeltaTimeMs = (weekDurationMs  + gpsSol.time - gpsData.lastNavSolTs) % weekDurationMs;`
to avoid the underflow before the modulo.